ystia/yorc
# Feature/scripts output

## Description of the change

### What I did

This pull request is just refactoring code, it doesn't provide any fix or new functionality from the user perspective.

The code refactored is the code managing logs of scripts/ansible playbooks, with the objective to have:
  * an open source implementation of the orchestrator providing :
	* logs of ansible playbook tasks once the task is finished, without having to wait for the end of the playbook
    * logs of a script output once the script is finished,
  * a premium implementation of the orchestrator in a separate private repository providing :
	* logs of ansible playbook tasks once the task is finished, like the open source version
    * logs of a script output live the orchestrator, without having to wait for the end of the script, contrarily to the open source version.

### How I did it

Introduced an interface outputHandler in prov/ansible/execution.go with two functions :
  * start() to start handling the outputs of a command
  * stop() to stop handling the outputs of a command
  
The interface is implemented by two types :
  * scriptOutputHandler in new file prov/ansible/scripts_output_handler_oss.go built only in the open source version (thanks to the build instruction +build !premium at the beginning of the file), handling the output of a script
  * playbookOutputHandler in new file  prov/ansible/playbook_output_handler.go common to both the open source and premium version of the orchestrator, handling the output of ansible playbook tasks.
  
Both types rely on already existing functions in the open source version (logAnsibleOutputInConsul and logAnsibleOutputInConsulFromScript) to provide the same functionnality as in the previous milestone.

The premium repository will provide a enhanced implementation of scriptOutputHandler to provide a script output in live.

Factorized as well some code with the aim of sharing it with the premium implementation.

### How to verify it

To verify there was no change introduced by this pull request, use a sample with both scripts and ansible playbooks to check logs are reported as in previous milestone :
  * a script standard output/error is provided in logs once its execution is finished
  * ansible playbook logs provide logs of a task as soon as the task is finished, without having to wait for the end of the playbook
 
In Alien4Cloud, menu Catalog, Manage Archives, Git Import, import the following sample :
  * Repository URL: https://github.com/laurentganne/tosca-samples.git
  * Branch: master
  * Archive to import: iterations
 
Create an application using the topology template ShortAndLongStartup
Run the application and check logs of each script/ansible playbook execution.
The bash component with a long startup is printing on stdout every 5 seconds for 12 iterations, and every 4 iterations is also printing on stderr
